### PR TITLE
Fix for edge case where an inactive User shadows the primary username.

### PR DESCRIPTION
**Pull request by eykd**
When the selected primary user tries to save itself with the primary username, a database error results. The simplest solution is to delete the inactive user, if it should exist. It should be safe to delete an inactive user, as the user was either never activated originally, or was deactivated as a result of a merge.

### CODE REVIEW

File: cas_consumer/backends.py [hunk 1]
```diff
@@ -247,20 +247,17 @@ def authenticate(self, ticket, service):
                 user.email = valid.attributes['email']
                 changed = True
 
-            if (self.set_username
-                and user.username != primary
-                ):
-                # Edge case: an inactive user with the primary username may
-                # exist, and it wasn't picked up before. We can neutralize it by
-                # generating a new username.
-                for shadow in User.objects.filter(username=primary, is_active=False):
```
Comment (eykd) [inline cas_consumer/backends.py]: We actually tried to account for this edge case originally, but alas, too late. The backend has already tried to save a new user w/ a pre-existing username, and failed w/ an exception, so this code is never reached. It's not really a good way to handle the edge case, anyway. Much better to re-activate the existing user.

**Comment by ryanmrubin** [comment]
FWIW, I've not looked at this before, nor do I really know how to test it, but this all makes sense to me from the internal clues and your description of everything. :+1:

File: cas_consumer/backends.py (hunk 1 above)
Comment (dwickwire) [inline cas_consumer/backends.py]: I sort of recall this, but what you say makes sense.  Let's do things we understand today rather than live in the "shadows" of the past :)

File: cas_consumer/backends.py [hunk 2]
```diff
@@ -247,20 +247,17 @@ def authenticate(self, ticket, service):
...
+            if (
+                    self.set_username
+                    and user.username != primary):
                 # Make the username match.
                 user.username = primary
                 changed = True
 
+            if not user.is_active:
+                user.is_active = True
+                changed = True
```
Comment (dwickwire) [inline cas_consumer/backends.py]: I wonder about the utility of tracking a state here.  Does it really help us, or, is there a some reason we're not saving in some cases?  Seems like we can maybe get rid of the changed state, but its been awhile since I've thought about this, and I'm only looking at this code in context of these 20 lines here.

Comment (eykd) [inline cas_consumer/backends.py]: This is necessary because the query above can now potentially pick up inactive users. However the user became inactive, we now need it to be active for Django auth's purposes, since this is now going to be the user associated with the CAS username.

**Comment by dwickwire** [comment]
:+1:

**Comment by eykd** [comment]
Thanks, guys.
